File: trifecta-ios/Trifecta/Core/Models/Thread.swift
import Foundation

enum RuntimeMode: String, Codable, Sendable {
    case approvalRequired = "approval-required"
    case autoAcceptEdits  = "auto-accept-edits"
    case fullAccess       = "full-access"
}

enum ProviderInteractionMode: String, Codable, Sendable {
    case `default`, plan
}

enum LatestTurnState: String, Codable, Sendable {
    case running, interrupted, completed, error
}

struct LatestTurn: Codable, Hashable, Sendable {
    let turnId: TurnID
    let state: LatestTurnState
    let requestedAt: Date
    let startedAt: Date?
    let completedAt: Date?
    let assistantMessageId: MessageID?

    private enum CodingKeys: String, CodingKey {
        case turnId, state, requestedAt, startedAt, completedAt, assistantMessageId
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        turnId = try c.decode(TurnID.self, forKey: .turnId)
        state = try c.decode(LatestTurnState.self, forKey: .state)
        requestedAt = try ISO8601Decoder.decodeDate(c, key: .requestedAt)
        startedAt = try (c.decodeIfPresent(String.self, forKey: .startedAt)).flatMap(ISO8601Decoder.parse)
        completedAt = try (c.decodeIfPresent(String.self, forKey: .completedAt)).flatMap(ISO8601Decoder.parse)
        assistantMessageId = try c.decodeIfPresent(MessageID.self, forKey: .assistantMessageId)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(turnId, forKey: .turnId)
        try c.encode(state, forKey: .state)
        try c.encode(ISO8601Decoder.formatter.string(from: requestedAt), forKey: .requestedAt)
        if let startedAt { try c.encode(ISO8601Decoder.formatter.string(from: startedAt), forKey: .startedAt) }
        if let completedAt { try c.encode(ISO8601Decoder.formatter.string(from: completedAt), forKey: .completedAt) }
        try c.encodeIfPresent(assistantMessageId, forKey: .assistantMessageId)
    }
}

enum SessionStatus: String, Codable, Sendable {
    case idle, starting, running, ready, interrupted, stopped, error
}

struct OrchestrationSession: Codable, Hashable, Sendable {
    let threadId: ThreadID
    let status: SessionStatus
    let providerName: String?
    let providerInstanceId: ProviderInstanceID?
    let runtimeMode: RuntimeMode
    let activeTurnId: TurnID?
    let lastError: String?
    let updatedAt: Date

    private enum CodingKeys: String, CodingKey {
        case threadId, status, providerName, providerInstanceId, runtimeMode, activeTurnId, lastError, updatedAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        threadId = try c.decode(ThreadID.self, forKey: .threadId)
        status = try c.decode(SessionStatus.self, forKey: .status)
        providerName = try c.decodeIfPresent(String.self, forKey: .providerName)
        providerInstanceId = try c.decodeIfPresent(ProviderInstanceID.self, forKey: .providerInstanceId)
        runtimeMode = (try? c.decode(RuntimeMode.self, forKey: .runtimeMode)) ?? .fullAccess
        activeTurnId = try c.decodeIfPresent(TurnID.self, forKey: .activeTurnId)
        lastError = try c.decodeIfPresent(String.self, forKey: .lastError)
        updatedAt = try ISO8601Decoder.decodeDate(c, key: .updatedAt)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(threadId, forKey: .threadId)
        try c.encode(status, forKey: .status)
        try c.encodeIfPresent(providerName, forKey: .providerName)
        try c.encodeIfPresent(providerInstanceId, forKey: .providerInstanceId)
        try c.encode(runtimeMode, forKey: .runtimeMode)
        try c.encodeIfPresent(activeTurnId, forKey: .activeTurnId)
        try c.encodeIfPresent(lastError, forKey: .lastError)
        try c.encode(ISO8601Decoder.formatter.string(from: updatedAt), forKey: .updatedAt)
    }
}

struct ThreadShell: Codable, Hashable, Sendable, Identifiable {
    let id: ThreadID
    let projectId: ProjectID
    var title: String
    var modelSelection: ModelSelection
    var runtimeMode: RuntimeMode
    var interactionMode: ProviderInteractionMode
    var branch: String?
    var worktreePath: String?
    var latestTurn: LatestTurn?
    let createdAt: Date
    var updatedAt: Date
    var archivedAt: Date?
    var session: OrchestrationSession?
    var latestUserMessageAt: Date?
    var hasPendingApprovals: Bool
    var hasPendingUserInput: Bool
    var hasActionableProposedPlan: Bool

    private enum CodingKeys: String, CodingKey {
        case id, projectId, title, modelSelection, runtimeMode, interactionMode, branch,
             worktreePath, latestTurn, createdAt, updatedAt, archivedAt, session,
             latestUserMessageAt, hasPendingApprovals, hasPendingUserInput, hasActionableProposedPlan
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(ThreadID.self, forKey: .id)
        projectId = try c.decode(ProjectID.self, forKey: .projectId)
        title = try c.decode(String.self, forKey: .title)
        modelSelection = try c.decode(ModelSelection.self, forKey: .modelSelection)
        runtimeMode = try c.decode(RuntimeMode.self, forKey: .runtimeMode)
        interactionMode = (try? c.decode(ProviderInteractionMode.self, forKey: .interactionMode)) ?? .default
        branch = try c.decodeIfPresent(String.self, forKey: .branch)
        worktreePath = try c.decodeIfPresent(String.self, forKey: .worktreePath)
        latestTurn = try c.decodeIfPresent(LatestTurn.self, forKey: .latestTurn)
        createdAt = try ISO8601Decoder.decodeDate(c, key: .createdAt)
        updatedAt = try ISO8601Decoder.decodeDate(c, key: .updatedAt)
        archivedAt = (try c.decodeIfPresent(String.self, forKey: .archivedAt)).flatMap(ISO8601Decoder.parse)
        session = try c.decodeIfPresent(OrchestrationSession.self, forKey: .session)
        latestUserMessageAt = (try c.decodeIfPresent(String.self, forKey: .latestUserMessageAt))
            .flatMap(ISO8601Decoder.parse)
        hasPendingApprovals = (try? c.decode(Bool.self, forKey: .hasPendingApprovals)) ?? false
        hasPendingUserInput = (try? c.decode(Bool.self, forKey: .hasPendingUserInput)) ?? false
        hasActionableProposedPlan = (try? c.decode(Bool.self, forKey: .hasActionableProposedPlan)) ?? false
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(projectId, forKey: .projectId)
        try c.encode(title, forKey: .title)
        try c.encode(modelSelection, forKey: .modelSelection)
        try c.encode(runtimeMode, forKey: .runtimeMode)
        try c.encode(interactionMode, forKey: .interactionMode)
        try c.encodeIfPresent(branch, forKey: .branch)
        try c.encodeIfPresent(worktreePath, forKey: .worktreePath)
        try c.encodeIfPresent(latestTurn, forKey: .latestTurn)
        try c.encode(ISO8601Decoder.formatter.string(from: createdAt), forKey: .createdAt)
        try c.encode(ISO8601Decoder.formatter.string(from: updatedAt), forKey: .updatedAt)
        if let archivedAt { try c.encode(ISO8601Decoder.formatter.string(from: archivedAt), forKey: .archivedAt) }
        try c.encodeIfPresent(session, forKey: .session)
        if let latestUserMessageAt {
            try c.encode(ISO8601Decoder.formatter.string(from: latestUserMessageAt), forKey: .latestUserMessageAt)
        }
        try c.encode(hasPendingApprovals, forKey: .hasPendingApprovals)
        try c.encode(hasPendingUserInput, forKey: .hasPendingUserInput)
        try c.encode(hasActionableProposedPlan, forKey: .hasActionableProposedPlan)
    }
}

struct ThreadDetail: Hashable, Sendable, Identifiable {
    let id: ThreadID
    let projectId: ProjectID
    var title: String
    var modelSelection: ModelSelection
    var runtimeMode: RuntimeMode
    var interactionMode: ProviderInteractionMode
    var branch: String?
    var worktreePath: String?
    var latestTurn: LatestTurn?
    let createdAt: Date
    var updatedAt: Date
    var archivedAt: Date?
    var messages: [Message]
    var session: OrchestrationSession?
    var proposedPlans: [ProposedPlan]
    var activities: [ThreadActivity]
}

extension ThreadDetail {
    nonisolated static func decode(from any: Any) throws -> ThreadDetail {
        let data = try JSONSerialization.data(withJSONObject: any)
        guard let dict = any as? [String: Any] else {
            throw DecodingError.dataCorrupted(.init(codingPath: [],
                debugDescription: "ThreadDetail expected an object"))
        }
        let baseDict = dict.filter { key, _ in
            !["proposedPlans", "activities"].contains(key)
        }
        let baseData = try JSONSerialization.data(withJSONObject: baseDict)
        let decoder = JSONDecoder()
        let base = try decoder.decode(BaseThreadDetail.self, from: baseData)
        _ = data

        let proposedPlans: [ProposedPlan] = (dict["proposedPlans"] as? [[String: Any]] ?? [])
            .compactMap { ProposedPlan.decode(from: $0) }
        let activities: [ThreadActivity] = (dict["activities"] as? [[String: Any]] ?? [])
            .compactMap { ThreadActivity.decode(from: $0) }

        return ThreadDetail(
            id: base.id,
            projectId: base.projectId,
            title: base.title,
            modelSelection: base.modelSelection,
            runtimeMode: base.runtimeMode,
            interactionMode: base.interactionMode,
            branch: base.branch,
            worktreePath: base.worktreePath,
            latestTurn: base.latestTurn,
            createdAt: base.createdAt,
            updatedAt: base.updatedAt,
            archivedAt: base.archivedAt,
            messages: base.messages.sorted { $0.createdAt < $1.createdAt },
            session: base.session,
            proposedPlans: proposedPlans,
            activities: activities
        )
    }
}

private struct BaseThreadDetail: Decodable {
    let id: ThreadID
    let projectId: ProjectID
    let title: String
    let modelSelection: ModelSelection
    let runtimeMode: RuntimeMode
    let interactionMode: ProviderInteractionMode
    let branch: String?
    let worktreePath: String?
    let latestTurn: LatestTurn?
    let createdAt: Date
    let updatedAt: Date
    let archivedAt: Date?
    let messages: [Message]
    let session: OrchestrationSession?

    private enum CodingKeys: String, CodingKey {
        case id, projectId, title, modelSelection, runtimeMode, interactionMode, branch,
             worktreePath, latestTurn, createdAt, updatedAt, archivedAt, messages, session
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(ThreadID.self, forKey: .id)
        projectId = try c.decode(ProjectID.self, forKey: .projectId)
        title = try c.decode(String.self, forKey: .title)
        modelSelection = try c.decode(ModelSelection.self, forKey: .modelSelection)
        runtimeMode = try c.decode(RuntimeMode.self, forKey: .runtimeMode)
        interactionMode = (try? c.decode(ProviderInteractionMode.self, forKey: .interactionMode)) ?? .default
        branch = try c.decodeIfPresent(String.self, forKey: .branch)
        worktreePath = try c.decodeIfPresent(String.self, forKey: .worktreePath)
        latestTurn = try c.decodeIfPresent(LatestTurn.self, forKey: .latestTurn)
        createdAt = try ISO8601Decoder.decodeDate(c, key: .createdAt)
        updatedAt = try ISO8601Decoder.decodeDate(c, key: .updatedAt)
        archivedAt = (try c.decodeIfPresent(String.self, forKey: .archivedAt)).flatMap(ISO8601Decoder.parse)
        messages = (try? c.decode([Message].self, forKey: .messages)) ?? []
        session = try c.decodeIfPresent(OrchestrationSession.self, forKey: .session)
    }
}

struct ProjectShell: Codable, Hashable, Sendable, Identifiable {
    let id: ProjectID
    var title: String
    var workspaceRoot: String
    var defaultModelSelection: ModelSelection?
    let createdAt: Date
    var updatedAt: Date

    private enum CodingKeys: String, CodingKey {
        case id, title, workspaceRoot, defaultModelSelection, createdAt, updatedAt
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(ProjectID.self, forKey: .id)
        title = try c.decode(String.self, forKey: .title)
        workspaceRoot = try c.decode(String.self, forKey: .workspaceRoot)
        defaultModelSelection = try c.decodeIfPresent(ModelSelection.self, forKey: .defaultModelSelection)
        createdAt = try ISO8601Decoder.decodeDate(c, key: .createdAt)
        updatedAt = try ISO8601Decoder.decodeDate(c, key: .updatedAt)
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(title, forKey: .title)
        try c.encode(workspaceRoot, forKey: .workspaceRoot)
        try c.encodeIfPresent(defaultModelSelection, forKey: .defaultModelSelection)
        try c.encode(ISO8601Decoder.formatter.string(from: createdAt), forKey: .createdAt)
        try c.encode(ISO8601Decoder.formatter.string(from: updatedAt), forKey: .updatedAt)
    }
}

// MARK: - Git / VCS

struct VcsWorkingTreeFile: Hashable, Sendable {
    let path: String
    let insertions: Int
    let deletions: Int
}

struct VcsWorkingTreeSummary: Hashable, Sendable {
    let files: [VcsWorkingTreeFile]
    let insertions: Int
    let deletions: Int
}

struct VcsStatusSummary: Hashable, Sendable {
    let isRepo: Bool
    let refName: String?
    let hasWorkingTreeChanges: Bool
    let hasUpstream: Bool
    let aheadCount: Int
    let behindCount: Int
    let workingTree: VcsWorkingTreeSummary

    nonisolated static func decode(from any: Any) throws -> VcsStatusSummary {
        let data = try JSONSerialization.data(withJSONObject: any)
        let root = try JSONDecoder().decode(VcsStatusSummaryPayload.self, from: data)
        return VcsStatusSummary(
            isRepo: root.isRepo,
            refName: root.refName,
            hasWorkingTreeChanges: root.hasWorkingTreeChanges,
            hasUpstream: root.hasUpstream,
            aheadCount: root.aheadCount,
            behindCount: root.behindCount,
            workingTree: VcsWorkingTreeSummary(
                files: root.workingTree.files.map {
                    VcsWorkingTreeFile(path: $0.path, insertions: $0.insertions, deletions: $0.deletions)
                },
                insertions: root.workingTree.insertions,
                deletions: root.workingTree.deletions
            )
        )
    }
}

private struct VcsStatusSummaryPayload: Decodable {
    struct WorkingTreePayload: Decodable {
        struct FilePayload: Decodable {
            let path: String
            let insertions: Int
            let deletions: Int
        }
        let files: [FilePayload]
        let insertions: Int
        let deletions: Int
    }

    let isRepo: Bool
    let refName: String?
    let hasWorkingTreeChanges: Bool
    let hasUpstream: Bool
    let aheadCount: Int
    let behindCount: Int
    let workingTree: WorkingTreePayload
}

struct VcsPullSummary: Hashable, Sendable {
    let status: String
    let refName: String
    let upstreamRef: String?

    nonisolated static func decode(from any: Any) throws -> VcsPullSummary {
        let data = try JSONSerialization.data(withJSONObject: any)
        let payload = try JSONDecoder().decode(VcsPullSummaryPayload.self, from: data)
        return VcsPullSummary(status: payload.status, refName: payload.refName, upstreamRef: payload.upstreamRef)
    }
}

private struct VcsPullSummaryPayload: Decodable {
    let status: String
    let refName: String
    let upstreamRef: String?
}

enum GitStackedAction: String, Sendable {
    case commit
    case push
}

struct GitActionSummary: Hashable, Sendable {
    let toastTitle: String
    let toastDescription: String?
}

File: trifecta-android/app/src/main/java/com/belweave/trifecta/core/models/Thread.kt
package com.belweave.trifecta.core.models

import kotlinx.serialization.json.JsonObject
import java.time.Instant

enum class RuntimeMode(val raw: String) {
    APPROVAL_REQUIRED("approval-required"),
    AUTO_ACCEPT_EDITS("auto-accept-edits"),
    FULL_ACCESS("full-access");

    companion object {
        fun fromRaw(raw: String?): RuntimeMode =
            values().firstOrNull { it.raw == raw } ?: FULL_ACCESS
    }
}

enum class ProviderInteractionMode(val raw: String) {
    DEFAULT("default"),
    PLAN("plan");

    companion object {
        fun fromRaw(raw: String?): ProviderInteractionMode =
            values().firstOrNull { it.raw == raw } ?: DEFAULT
    }
}

enum class LatestTurnState(val raw: String) {
    RUNNING("running"),
    INTERRUPTED("interrupted"),
    COMPLETED("completed"),
    ERROR("error");

    companion object {
        fun fromRaw(raw: String?): LatestTurnState =
            values().firstOrNull { it.raw == raw } ?: COMPLETED
    }
}

data class LatestTurn(
    val turnId: TurnID,
    val state: LatestTurnState,
    val requestedAt: Instant,
    val startedAt: Instant?,
    val completedAt: Instant?,
    val assistantMessageId: MessageID?
) {
    companion object {
        fun fromJson(obj: JsonObject?): LatestTurn? {
            if (obj == null) return null
            val turnId = obj.str("turnId")?.let { TurnID(it) } ?: return null
            val state = LatestTurnState.fromRaw(obj.str("state"))
            val requestedAt = Iso8601.parse(obj.str("requestedAt")) ?: Instant.now()
            return LatestTurn(
                turnId = turnId,
                state = state,
                requestedAt = requestedAt,
                startedAt = Iso8601.parse(obj.str("startedAt")),
                completedAt = Iso8601.parse(obj.str("completedAt")),
                assistantMessageId = obj.str("assistantMessageId")?.let { MessageID(it) }
            )
        }
    }
}

enum class SessionStatus(val raw: String) {
    IDLE("idle"),
    STARTING("starting"),
    RUNNING("running"),
    READY("ready"),
    INTERRUPTED("interrupted"),
    STOPPED("stopped"),
    ERROR("error");

    companion object {
        fun fromRaw(raw: String?): SessionStatus =
            values().firstOrNull { it.raw == raw } ?: IDLE
    }
}

data class OrchestrationSession(
    val threadId: ThreadID,
    val status: SessionStatus,
    val providerName: String?,
    val providerInstanceId: ProviderInstanceID?,
    val runtimeMode: RuntimeMode,
    val activeTurnId: TurnID?,
    val lastError: String?,
    val updatedAt: Instant
) {
    companion object {
        fun fromJson(obj: JsonObject?): OrchestrationSession? {
            if (obj == null) return null
            val threadId = obj.str("threadId")?.let { ThreadID(it) } ?: return null
            return OrchestrationSession(
                threadId = threadId,
                status = SessionStatus.fromRaw(obj.str("status")),
                providerName = obj.str("providerName"),
                providerInstanceId = obj.str("providerInstanceId")?.let { ProviderInstanceID(it) },
                runtimeMode = RuntimeMode.fromRaw(obj.str("runtimeMode")),
                activeTurnId = obj.str("activeTurnId")?.let { TurnID(it) },
                lastError = obj.str("lastError"),
                updatedAt = Iso8601.parse(obj.str("updatedAt")) ?: Instant.now()
            )
        }
    }
}

data class ThreadShell(
    val id: ThreadID,
    val projectId: ProjectID,
    val title: String,
    val modelSelection: ModelSelection,
    val runtimeMode: RuntimeMode,
    val interactionMode: ProviderInteractionMode,
    val branch: String?,
    val worktreePath: String?,
    val latestTurn: LatestTurn?,
    val createdAt: Instant,
    val updatedAt: Instant,
    val archivedAt: Instant?,
    val session: OrchestrationSession?,
    val latestUserMessageAt: Instant?,
    val hasPendingApprovals: Boolean,
    val hasPendingUserInput: Boolean,
    val hasActionableProposedPlan: Boolean
) {
    companion object {
        fun fromJson(obj: JsonObject): ThreadShell? {
            val id = obj.str("id")?.let { ThreadID(it) } ?: return null
            val projectId = obj.str("projectId")?.let { ProjectID(it) } ?: return null
            val title = obj.str("title") ?: "Untitled"
            val modelSelection = ModelSelection.fromJson(obj.obj("modelSelection")) ?: return null
            return ThreadShell(
                id = id,
                projectId = projectId,
                title = title,
                modelSelection = modelSelection,
                runtimeMode = RuntimeMode.fromRaw(obj.str("runtimeMode")),
                interactionMode = ProviderInteractionMode.fromRaw(obj.str("interactionMode")),
                branch = obj.str("branch"),
                worktreePath = obj.str("worktreePath"),
                latestTurn = LatestTurn.fromJson(obj.obj("latestTurn")),
                createdAt = Iso8601.parse(obj.str("createdAt")) ?: Instant.now(),
                updatedAt = Iso8601.parse(obj.str("updatedAt")) ?: Instant.now(),
                archivedAt = Iso8601.parse(obj.str("archivedAt")),
                session = OrchestrationSession.fromJson(obj.obj("session")),
                latestUserMessageAt = Iso8601.parse(obj.str("latestUserMessageAt")),
                hasPendingApprovals = obj.bool("hasPendingApprovals") ?: false,
                hasPendingUserInput = obj.bool("hasPendingUserInput") ?: false,
                hasActionableProposedPlan = obj.bool("hasActionableProposedPlan") ?: false
            )
        }
    }
}

data class ProjectShell(
    val id: ProjectID,
    val title: String,
    val workspaceRoot: String,
    val defaultModelSelection: ModelSelection?,
    val createdAt: Instant,
    val updatedAt: Instant
) {
    companion object {
        fun fromJson(obj: JsonObject): ProjectShell? {
            val id = obj.str("id")?.let { ProjectID(it) } ?: return null
            return ProjectShell(
                id = id,
                title = obj.str("title") ?: "Project",
                workspaceRoot = obj.str("workspaceRoot") ?: "",
                defaultModelSelection = ModelSelection.fromJson(obj.obj("defaultModelSelection")),
                createdAt = Iso8601.parse(obj.str("createdAt")) ?: Instant.now(),
                updatedAt = Iso8601.parse(obj.str("updatedAt")) ?: Instant.now()
            )
        }
    }
}

data class ThreadDetail(
    val id: ThreadID,
    val projectId: ProjectID,
    val title: String,
    val modelSelection: ModelSelection,
    val runtimeMode: RuntimeMode,
    val interactionMode: ProviderInteractionMode,
    val branch: String?,
    val worktreePath: String?,
    val latestTurn: LatestTurn?,
    val createdAt: Instant,
    val updatedAt: Instant,
    val archivedAt: Instant?,
    val messages: List<Message>,
    val session: OrchestrationSession?,
    val proposedPlans: List<ProposedPlan>,
    val activities: List<ThreadActivity>
) {
    companion object {
        fun fromJson(obj: JsonObject): ThreadDetail? {
            val id = obj.str("id")?.let { ThreadID(it) } ?: return null
            val projectId = obj.str("projectId")?.let { ProjectID(it) } ?: return null
            val title = obj.str("title") ?: "Untitled"
            val modelSelection = ModelSelection.fromJson(obj.obj("modelSelection")) ?: return null
            val messages = obj.arr("messages")
                ?.mapNotNull { Message.fromJson(it.asObjectOrNull() ?: return@mapNotNull null) }
                ?.sortedBy { it.createdAt }
                ?: emptyList()
            val plans = obj.arr("proposedPlans")
                ?.mapNotNull { ProposedPlan.fromJson(it.asObjectOrNull()) }
                ?: emptyList()
            val activities = obj.arr("activities")
                ?.mapNotNull { ThreadActivity.fromJson(it.asObjectOrNull()) }
                ?: emptyList()
            return ThreadDetail(
                id = id,
                projectId = projectId,
                title = title,
                modelSelection = modelSelection,
                runtimeMode = RuntimeMode.fromRaw(obj.str("runtimeMode")),
                interactionMode = ProviderInteractionMode.fromRaw(obj.str("interactionMode")),
                branch = obj.str("branch"),
                worktreePath = obj.str("worktreePath"),
                latestTurn = LatestTurn.fromJson(obj.obj("latestTurn")),
                createdAt = Iso8601.parse(obj.str("createdAt")) ?: Instant.now(),
                updatedAt = Iso8601.parse(obj.str("updatedAt")) ?: Instant.now(),
                archivedAt = Iso8601.parse(obj.str("archivedAt")),
                messages = messages,
                session = OrchestrationSession.fromJson(obj.obj("session")),
                proposedPlans = plans,
                activities = activities
            )
        }
    }
}

// MARK: - VCS

data class VcsWorkingTreeFile(val path: String, val insertions: Int, val deletions: Int)
data class VcsWorkingTreeSummary(
    val files: List<VcsWorkingTreeFile>,
    val insertions: Int,
    val deletions: Int
)

data class VcsStatusSummary(
    val isRepo: Boolean,
    val refName: String?,
    val hasWorkingTreeChanges: Boolean,
    val hasUpstream: Boolean,
    val aheadCount: Int,
    val behindCount: Int,
    val workingTree: VcsWorkingTreeSummary
) {
    companion object {
        fun fromJson(obj: JsonObject): VcsStatusSummary? {
            val workingTreeObj = obj.obj("workingTree") ?: return null
            val files = workingTreeObj.arr("files")?.mapNotNull { f ->
                val o = f.asObjectOrNull() ?: return@mapNotNull null
                VcsWorkingTreeFile(
                    path = o.str("path") ?: return@mapNotNull null,
                    insertions = o.intAt("insertions") ?: 0,
                    deletions = o.intAt("deletions") ?: 0
                )
            } ?: emptyList()
            return VcsStatusSummary(
                isRepo = obj.bool("isRepo") ?: false,
                refName = obj.str("refName"),
                hasWorkingTreeChanges = obj.bool("hasWorkingTreeChanges") ?: false,
                hasUpstream = obj.bool("hasUpstream") ?: false,
                aheadCount = obj.intAt("aheadCount") ?: 0,
                behindCount = obj.intAt("behindCount") ?: 0,
                workingTree = VcsWorkingTreeSummary(
                    files = files,
                    insertions = workingTreeObj.intAt("insertions") ?: 0,
                    deletions = workingTreeObj.intAt("deletions") ?: 0
                )
            )
        }
    }
}

data class VcsPullSummary(
    val status: String,
    val refName: String,
    val upstreamRef: String?
) {
    companion object {
        fun fromJson(obj: JsonObject): VcsPullSummary? {
            return VcsPullSummary(
                status = obj.str("status") ?: return null,
                refName = obj.str("refName") ?: return null,
                upstreamRef = obj.str("upstreamRef")
            )
        }
    }
}
